@feature/example 6 content to web

rohit-lakhanpal opened the pull request.

Commit:
Inital commit for example 6 - content to web

Initial commit for example 6. A simple .Net 5 and React Web Application that allows you to search and display panviva documents.

Commit:
Added Samples.Net.Web with PanvivaController

Added PanvivaController to the .NET 5 React Template Project

Commit:
Added client app to utilise Panviva APIs

Added a simple client app that is able to search and view documents.

Commit:
Updated Documentaion

Added documentation and screenshots.

Commit:
Code cleanup and User Experience Improvements

- Added ability to fetch files via the api
- Improved User Experience for the homepage
- Improved User Experience for search & search result list

Commit:
Updated screenshots for documentation

Updated screenshots depicting new user experience for Homepage & Search Results Page

Commit:
Updated screenshot filename to fix markdown references

Commit:
Code Cleanup and User Experience Improvements

- moved document to its own folder component
- updated screenshots
- UI changes to nav bar

Commit:
Fix missing dependancies

Missing Moment dependancy.

Commit:
Added error handling

- Added API error exception action filter on the server
- Added API error handling in UI

Commit:
Cleanup Document implementation

- Breakup document into sub components
- Added document styling css

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/Document.js:
```diff
@@ -0,0 +1,131 @@
+      ...panvivaDocument,
+      id: params.id,
+    });
+  }, [params]);
```
Add `panvivaDocument` to the dependancy array after `params.`

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/Document.js:
```diff
@@ -0,0 +1,131 @@
+    if (panvivaDocument && panvivaDocument.id) {
+      getDocument(panvivaDocument.id);
+    }
+  }, [panvivaDocument.id]);
```
Just adding `panvivaDocument` to the dependancy array is enough.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/Document.js:
```diff
@@ -0,0 +1,131 @@
+    document.title = panvivaDocument.loading
+      ? "Panviva - Loading ..."
+      : `Panviva - ${panvivaDocument?.document?.name} [#${panvivaDocument?.document?.id}]`;
+  }, [panvivaDocument.document]);
```
Add `panvivaDocument` and `document` to the dependancy array.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/Document.js:
```diff
@@ -0,0 +1,131 @@
+      element < htmlObject?.getElementsByTagName("img").length;
+      element++
+    ) {
+      debugger;
```
Leftover `debugger` statement.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/Document.js:
```diff
@@ -0,0 +1,131 @@
+      element++
+    ) {
+      debugger;
+      let imageId = htmlObject
```
Fix the indentation here from line 74 to line 79

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/DocumentSectionAndSubSectionBody.js:
```diff
@@ -0,0 +1,47 @@
+export const DocumentSectionAndSubSectionBody = (props) => {
+  const fixImages = (htmlObject) => {
+    let panvivaImages = htmlObject?.getElementsByTagName("img");
+    for (let idx = 0; idx < panvivaImages.length; idx++) {
```
use [Array.prototype.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) since you are not using the index for anything other that retrieving the object from the array.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/DocumentSectionAndSubSectionBody.js:
```diff
@@ -0,0 +1,47 @@
+  const fixTooltips = (htmlObject) => {
+    let tooltipClassName = "pv-tooltip-link";
+    let tooltipElements = htmlObject?.getElementsByClassName(tooltipClassName);
+    for (let idx = 0; idx < tooltipElements.length; idx++) {
```
use [Array.prototype.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) since you are not using the index for anything other that retrieving the object from the array.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/Search.js:
```diff
@@ -0,0 +1,82 @@
+  const [title, setTitle] = useState("");
+  const [pendingQuery, setPendingQuery] = useState("");
+  const [searchQuery, setSearchQuery] = useState(
+    params && params.query !== "" ? params.query : "*"
```
Can shorten this to `params?.query || "*"`

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/Search.js:
```diff
@@ -0,0 +1,82 @@
+  // update the document title
+  useEffect(() => {
+    document.title = title;
+  }, [title]);
```
dependancy should be `document`.

rohit-lakhanpal replied:
Setting the html document title.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/Error.js:
```diff
@@ -0,0 +1,27 @@
+      : "Undefined error."
+  );
+  const [status] = useState(
+    params && params.status && params.status !== "" ? params.status : "500"
```
can shorten to `params?.status || "500"`

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/Home.js:
```diff
@@ -0,0 +1,44 @@
+      >
+        <SearchBar
+          handleInputChange={(e) => {
+            setSearchQuery(e.target.value);
```
Events can sometimes have undefineds or nulls. Just to be safe add `e?.target?.value`

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/Layout.js:
```diff
@@ -0,0 +1,18 @@
+import { Container } from 'reactstrap';
+import { NavMenu } from './NavMenu';
+
+export class Layout extends Component {
```
Convert to a function component for consistency.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/NavMenu.js:
```diff
@@ -0,0 +1,67 @@
+import { Link } from "react-router-dom";
+import "./NavMenu.css";
+
+export class NavMenu extends Component {
```
Convert to a functional component for consistency.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/DocumentSection.js:
```diff
@@ -0,0 +1,21 @@
+      <h2 id={props.section.id}># {props.section.name}</h2>
+
+      {props?.subSecions.map((subSection) => {
+        return (
```
Don't need an explicit return statement.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/document/DocumentSubSection.js:
```diff
@@ -0,0 +1,11 @@
+import { DocumentSectionAndSubSectionBody } from "./DocumentSectionAndSubSectionBody";
+
+export const DocumentSubSection = (props) => {
+  return (
```
Don't need an explicit return statement.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/Search.js:
```diff
@@ -0,0 +1,82 @@
+      searchContent();
+      setTitle(`Panviva - Search everything!`);
+    }
+  }, [searchQuery]);
```
Add `params` to the dependancy array.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/Search.js:
```diff
@@ -0,0 +1,82 @@
+      let errorMessage = JSON.stringify(error);
+      let path = `/error/unknown/${errorMessage}`;
+      console.error(errorMessage);
+      console.error(errorMessage);
```
Unnecessary log statement.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchResultsListItem.js:
```diff
@@ -0,0 +1,74 @@
+
+  const getFirstDocumentUrl = () => {
+    if (
+      props &&
```
Can be shortened to `params?.searchResults?.results?.length > 0`

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchResultsList.js:
```diff
@@ -0,0 +1,30 @@
+import { SearchResultsListItem } from "./SearchResultsListItem";
+
+export const SearchResultsList = (props) => {
+  return (
```
No need for explicit `return`.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchResultsHeading.js:
```diff
@@ -0,0 +1,24 @@
+import { ProgressBar } from "react-bootstrap";
+
+export const SearchResultsHeading = (props) => {
+  return (
```
No need for explicit `return`.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchResults.js:
```diff
@@ -0,0 +1,20 @@
+
+export const SearchResults = (props) => {
+  //isLoading
+  return (
```
No need for explicit `return`.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchBar.js:
```diff
@@ -0,0 +1,37 @@
+import { Search } from "react-bootstrap-icons";
+
+export const SearchBar = (props) => {
+  return (
```
No need for explicit `return`.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/App.js:
```diff
@@ -0,0 +1,24 @@
+
+import "./custom.css";
+
+export default class App extends Component {
```
Convert to functional component for consistency.

r3core reviewed (commented):
Just some overall thoughts.

- I wasn't considering HTML manipulation in this review. I assume you didn't find a good library?
- Usually it's good to use the `.jsx` suffix for filenames for files with React code inside them.
- CSS files + class names inside JSX components is fine. However we should be looking at styled components.
- A lot of error handling code is present within the controller that does nothing but logging. Wouldn't it be better to move to the exception filter?

Commit:
Code Review - Move to functional component

Moved template code to functional components for consistency:
- App
- Layout
- NavMenu

Commit:
CSS Updates - Show task headers

Show task headers rendered by Panviva APIs

Commit:
Search fix search result item click behaviour

- Using onClick to trigger react router instead of href causing reloads.

Commit:
Removing custom headings as they're generated via Panviva APIs

Commit:
Added scrolling capibility

Added capability to scroll to sections/subsections.

Commit:
Update document.png

Updated document screenshot

Commit:
Code Review Fixes

- Adding optional chaining to simplify

Commit:
Code Improvement

- Using array map instead of for each
- Making error messages more explicit
- Remove unused dependencies
- Fix linting warnings

Commit:
Update Home.js

- Fix possible null exception

Commit:
Update Search.js

Removed duplicate error message.

r3core commented on Samples.Net.Web/Samples.Net.Web/ClientApp/src/components/search/SearchResults.js:
```diff
@@ -0,0 +1,20 @@
+import { SearchResultsList } from "./SearchResultsList";
+
+export const SearchResults = (props) => {
+  //isLoading
```
Remove comment.

r3core commented on Samples.Net.Web/Samples.Net.Web/appsettings.Development.json:
```diff
@@ -0,0 +1,9 @@
+      "Default": "Information",
+      "Microsoft": "Warning",
+      "Microsoft.Hosting.Lifetime": "Information"
+    }
```
Add the Panviva settings here as well.

Commit:
Code review fixes

- Added config placeholders to dev app settings
- Removed unnecessary comments
